### PR TITLE
Bookstore implemented

### PR DESCRIPTION
**Pull request by WoLkErSs**

### CODE REVIEW

File: app/mailers/application_mailer.rb [hunk 1]
```diff
@@ -1,4 +1,4 @@
 class ApplicationMailer < ActionMailer::Base
-  default from: 'from@example.com'
+  default from: ENV['GMAIL_SMTP_PASSWORD']
```
Comment (za4emnik) [inline app/mailers/application_mailer.rb]: email?

File: app/mailers/order_mailer.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class OrderMailer < ApplicationMailer
+  default from: ENV['GMAIL_SMTP_PASSWORD'],
```
Comment (za4emnik) [inline app/mailers/order_mailer.rb]: ?

File: app/models/delivery_service.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+class DeliveryService < ApplicationRecord
```
Comment (za4emnik) [inline app/models/delivery_service.rb]: remove

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable,
+         :confirmable, :omniauthable, omniauth_providers: [:facebook]
+
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).or(where(email: auth.info.email)).first_or_create do |user|
+      user.provider = auth.provider
+      user.uid = auth.uid
+      user.email = auth.info.email
+      user.password = 'Wdw12ESee3'
```
Comment (za4emnik) [inline app/models/user.rb]: frendly token device

File: app/presenters/quick_registration_presenter.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class QuickRegistrationPresenter
+  def checkout_step
+    'quick_registration'
```
Comment (za4emnik) [inline app/presenters/quick_registration_presenter.rb]: const

File: app/services/setup_filter_service.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class SetupFilterService
+  def call(filter)
+    if BooksQuery::FILTERS.values.include?(filter)
```
Comment (za4emnik) [inline app/services/setup_filter_service.rb]: one line

File: spec/controllers/checkout_steps_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+      it 'render payment template' do
+        order.update(checkout: 'payment')
+        subject
+        expect(assigns(:presenter)).to be_kind_of(PaymentPresenter)
+        expect(subject).to render_template(:payment)
+      end
+
+      it 'render confirm template' do
+        order.update(checkout: 'confirm')
```
Comment (za4emnik) [inline spec/controllers/checkout_steps_controller_spec.rb]: confirm!

File: spec/controllers/checkout_steps_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        subject
+        expect(assigns(:presenter)).to be_kind_of(ConfirmPresenter)
+        expect(subject).to render_template(:confirm)
+      end
+
+      it 'render complete template' do
+        order.update(checkout: 'complete')
```
Comment (za4emnik) [inline spec/controllers/checkout_steps_controller_spec.rb]: checkout.complete!

File: spec/controllers/checkout_steps_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        it 'address template' do
+          expect(subject).to render_template(:address)
+        end
+      end
+
+      context 'when fill correct data' do
+        let(:params) { order_params }
+
+        context 'when filling billing and shipping address' do
+          let(:order_params) { { 'order': { 'billing_address': billing_address,
```
Comment (za4emnik) [inline spec/controllers/checkout_steps_controller_spec.rb]: symbols

File: spec/controllers/order_items_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,174 @@
+RSpec.describe OrderItemsController, type: :controller do
+  describe '.update' do
+    subject{ patch :update, params: params }
+    context 'when user is guest' do
+      let(:quantity) { '1' }
+      let(:book_id) { '1' }
```
Comment (za4emnik) [inline spec/controllers/order_items_controller_spec.rb]: int

File: spec/controllers/order_items_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+      let(:params) { { 'book_id' => book_id, 'value' => quantity } }
```
Comment (za4emnik) [inline spec/controllers/order_items_controller_spec.rb]: symbols

File: spec/controllers/orders_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,17 @@
+RSpec.describe OrdersController, typr: :controller do
+
+  describe '#index' do
+    let!(:user) { create(:user, :with_order) }
+    let!(:params) { { "filter"=>"In progress" } }
```
Comment (za4emnik) [inline spec/controllers/orders_controller_spec.rb]: symbols

File: spec/controllers/registrations_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,65 @@
+RSpec.describe Users::RegistrationsController, type: :controller do
+  describe '#create' do
+    let(:password) { SecureRandom.base64(10) }
+    let(:password_confirmation) { password }
+    let(:params) { { 'user': { 'email': email,
```
Comment (za4emnik) [inline spec/controllers/registrations_controller_spec.rb]: use factory

File: spec/controllers/registrations_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        expect { subject }.to change { User.all.count }.by(1)
+      end
+    end
+
+    context 'incorrect email' do
+      let(:email) { '@.com' }
+
+      it 'redirect with notice' do
+        subject
+        expect(flash[:notice]).to eq('You can`t use this email')
```
Comment (za4emnik) [inline spec/controllers/registrations_controller_spec.rb]: I18n

File: spec/controllers/reviews_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+  describe '.create' do
+    let(:book) { create(:book) }
+    let(:user) { create(:user) }
+    let(:title) { FFaker::Lorem.characters(10) }
+    let(:body) { FFaker::Lorem.characters(50) }
+    let(:score) { rand(1..5) }
+
+
+    context 'when save review' do
+      let(:params) { { 'book_id': book.id, 'user_id': user.id, 'score': score, 'title': title, 'body': body } }
```
Comment (za4emnik) [inline spec/controllers/reviews_controller_spec.rb]: attributes_for

File: spec/controllers/users_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        end
+      end
+    end
+  end
+
+  describe '.update_password' do
+
+    subject(:controller_action) { patch 'update', params: params }
+
+    let(:params) do
```
Comment (za4emnik) [inline spec/controllers/users_controller_spec.rb]: attributes_for

File: spec/controllers/users_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+        end
+
+        it 'redirect to settings' do
+          expect(controller).to redirect_to(settings_path)
+        end
+      end
+    end
+
+    context 'when passwords match' do
+      let(:password) { 'Dosenotma4' }
```
Comment (za4emnik) [inline spec/controllers/users_controller_spec.rb]: equals lets

File: spec/controllers/users_controller_spec.rb [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+      end
+
+      it 'update billing attributes ' do
+        expect { controller_action }.to change { controller.current_user.billing_address }
+      end
+    end
+  end
+
+  describe '.update_shipping' do
+    subject(:controller_action) { patch 'update', params: params }
```
Comment (za4emnik) [inline spec/controllers/users_controller_spec.rb]: symbols

File: spec/decorators/book_decorator_spec.rb [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+  it 'returns the displayed days' do
+    expect(book.stars).to eq([5, 4, 3, 2, 1])
+  end
+
+  it 'returns the displayed days' do
+    expect(book.reviews_quantity).to eq(0)
+  end
+
+  it 'returns the displayed days' do
+    expect(book.authors_all).to eq(book.authors.map { |a| a.first_name + ' ' + a.last_name }.join(', '))
```
Comment (za4emnik) [inline spec/decorators/book_decorator_spec.rb]: `a.first_name + ' ' + a.last_name` replase to string

File: spec/factories/addresses.rb [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+    trait :billing_address do
+      type { 'BillingAddress' }
+    end
+
+    trait :shipping_address do
+      type { 'ShippingAddress' }
+    end
+
+    trait :correct_address do
+      first_name { 'Pupa' }
```
Comment (za4emnik) [inline spec/factories/addresses.rb]: FFacker

File: spec/factories/credit_cards.rb [hunk 21]
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
+  factory :credit_card do
+    association :order
+    number { 16.times.map{rand(10)}.join }
+    cvv { FFaker::PhoneNumber.area_code }
+    name { 'Universal' }
```
Comment (za4emnik) [inline spec/factories/credit_cards.rb]: FFacker

File: spec/factories/orders.rb [hunk 22]
```diff
@@ -0,0 +1,21 @@
+FactoryBot.define do
+  factory :order do
+    association :user
+    association :delivery_service
+    total { 10 }
+    status { 'in_progress' }
```
Comment (za4emnik) [inline spec/factories/orders.rb]: statuses `in_progress`

File: spec/features/checkout_steps/confirm_spec.rb [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+    user.orders.first.to_delivery!
+    user.orders.first.to_payment!
+    user.orders.first.to_confirm!
+    login_as user, scope: :user
+  end
+
+  it 'user move to change address' do
+    visit '/checkout_step'
+    click_link('shipping')
+    expect(page).to have_content 'Billing Address'
```
Comment (za4emnik) [inline spec/features/checkout_steps/confirm_spec.rb]: I18n

File: spec/models/books_author_spec.rb [hunk 24]
```diff
@@ -0,0 +1,2 @@
+RSpec.describe BooksAuthor, type: :model do
```
Comment (za4emnik) [inline spec/models/books_author_spec.rb]: remove

File: spec/presenters/short_address_presenter_spec.rb [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+  let!(:order_billing_address) { create(:order_billing_address, addressable: order) }
+  let!(:order_shipping_address) { create(:order_shipping_address, addressable: order) }
+
+  before do
+    allow(subject.instance_variable_set(:@order, order))
+  end
+
+  describe '#full_address' do
+    context 'when show full shipping and billing address' do
+      it { expect(subject.full_address('billing_address')).to eq(order.billing_address.first_name + ' ' + order.billing_address.last_name) }
```
Comment (za4emnik) [inline spec/presenters/short_address_presenter_spec.rb]: `order.billing_address.first_name + ' ' + order.billing_address.last_name` to string

File: spec/queries/books_query_spec.rb [hunk 26]
```diff
@@ -0,0 +1,41 @@
+require 'rails_helper'
+
+RSpec.describe BooksQuery do
+  subject(:query) { described_class }
+
+  let(:book_a) { create(:book, title: 'Aaaaa', price: 3, sold: 1) }
```
Comment (za4emnik) [inline spec/queries/books_query_spec.rb]: refactor

File: spec/queries/books_query_spec.rb [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+    it 'Z-A sort' do
+      expect(query.new(Book.all).books_sort('Title Z-A')).to eq([book_c, book_b, book_a])
+    end
+
+    it 'Sort price: Low to high' do
+      expect(query.new(Book.all).books_sort('Price: Low to high')).to eq([book_c, book_b, book_a])
+    end
+
+    it 'Sort price: High to low' do
+      expect(query.new(Book.all).books_sort('Price: High to low')).to eq([book_a, book_b, book_c])
```
Comment (za4emnik) [inline spec/queries/books_query_spec.rb]: I18n

File: spec/rails_helper.rb [hunk 28]
```diff
@@ -0,0 +1,60 @@
+require 'spec_helper'
+ENV['RAILS_ENV'] ||= 'test'
+require 'simplecov'
+SimpleCov.start do
```
Comment (za4emnik) [inline spec/rails_helper.rb]: to support

File: spec/services/items_quantity_service_spec.rb [hunk 29]
```diff
@@ -0,0 +1,33 @@
+require 'rails_helper'
+
+RSpec.describe ItemsQuantityService do
+  let(:guest_items) { {'1' => '1'} }
```
Comment (za4emnik) [inline spec/services/items_quantity_service_spec.rb]: symbol

File: spec/services/summary_service_spec.rb [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+      expect( service.call(nil, guest_items, coupon.id, nil) ).to eq({coupon: coupon.value, subtotal: 0, total: 0})
+    end
+  end
+
+  context 'when apply coupon which less when subtotal' do
+    let(:guest_items) { { book.id.to_s => '100' } }
+    let(:subtotal_value) { (book.price * guest_items[book.id.to_s].to_i).round(2) }
+    let(:total_value) { (subtotal_value - coupon.value).round(2) }
+
+    it 'guest appla coupon' do
```
Comment (za4emnik) [inline spec/services/summary_service_spec.rb]: apply

File: spec/controllers/order_items_controller_spec.rb (hunk 10 above)
Comment (WoLkErSs) [reply spec/controllers/order_items_controller_spec.rb]: This variable for compare session

File: spec/controllers/registrations_controller_spec.rb (hunk 13 above)
Comment (WoLkErSs) [reply spec/controllers/registrations_controller_spec.rb]: Factory is inappropriate

File: app/forms/credit_card_form.rb [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+    if correct_expiry_date?
+      expiry = Date.parse("20#{expiry_date.last(2)}-#{expiry_date.first(2)}-01")
+      errors.add(:expiry_date, I18n.t('forms.expired_card_date')) unless Time.zone.now.strftime('%Y-%m-%d').to_date < expiry
+    else
+      errors.add(:expiry_date, I18n.t('forms.wrong_format_date_card'))
+    end
+  end
+
+  def correct_expiry_date?
+    permissible_digit?(expiry_date.first(2), 12) && permissible_digit?(expiry_date.last(2))
```
Comment (za4emnik) [inline app/forms/credit_card_form.rb]: rename

File: app/helpers/flashes_helper.rb [hunk 32]
```diff
@@ -0,0 +1,5 @@
+module FlashesHelper
+  def alert?(string)
+    string == I18n.t('flash.alert')
```
Comment (za4emnik) [inline app/helpers/flashes_helper.rb]: change

File: app/models/review.rb [hunk 33]
```diff
@@ -0,0 +1,25 @@
...
+    event :to_approve do
+      transitions from: :unapproved, to: :approved
+    end
+
+    event :to_unapprove do
+      transitions from: :approved, to: :unapproved
+    end
+  end
+
+  scope :unapproved, -> { where(status: 'unapproved') }
```
Comment (za4emnik) [inline app/models/review.rb]: check, please

File: app/presenters/confirm_presenter.rb [hunk 34]
```diff
@@ -0,0 +1,22 @@
...
+    @credit_card = credit_card
+    @coupon = coupon
+  end
+
+  def checkout_step
+    :confirm
+  end
+
+  def payment_info_card_number
+    '**** **** **** ' + @credit_card.number.last(4)
```
Comment (za4emnik) [inline app/presenters/confirm_presenter.rb]: change

File: app/services/update_checkout_service.rb [hunk 35]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def assign_credit_card(order, params)
+    credit_card = CreditCardForm.new(card_params(params))
+    change_order_checkout(order, VIEW_STEPS[:confirm]) if credit_card.save(order)
+    credit_card
+  end
+
+  def address_error?(new_address)
+    new_address.first.errors.empty? && new_address.last.errors.empty?
```
Comment (za4emnik) [inline app/services/update_checkout_service.rb]: present?

File: app/views/books/partials/_book.haml [hunk 36]
```diff
@@ -0,0 +1,58 @@
+.col-sm-6.clearfix
+  .product-gallery
+    .mb-20
+      = image_tag( book.image_url, alt: 'design-book', class: 'img-responsive')
+      .images.clearfix
+        %img.minimized{alt: 'design-book', src: 'https://gdurl.com/SjK8', class: 'img-link'}/
```
Comment (za4emnik) [inline app/views/books/partials/_book.haml]: ?

File: app/views/checkout_steps/quick_registration.haml [hunk 37]
```diff
@@ -0,0 +1,51 @@
+%main.container
+  .row
+    .col-md-5.mb-40
+      .text-center
+        %h3.general-subtitle
+          = t('quick_registration.login.title')
+        = link_to user_facebook_omniauth_authorize_path, class: 'general-login-icon' do
+          %i.fa.fa-facebook-official
+        %p.general-login-text or
+      = simple_form_for(User.new, defaults: { input_html: { class: 'general-form' }}, url: session_path(User.new)) do |f|
```
Comment (za4emnik) [inline app/views/checkout_steps/quick_registration.haml]: User.new?

File: app/views/checkout_steps/quick_registration.haml [hunk 38]
```diff
@@ -0,0 +1,51 @@
...
+              placeholder: 'Enter email',
+              required: true,
+              autofocus: true,
+              input_html: { autocomplete: 'email' }, class: 'form-control'
+          .form-group.mb-30
+            = f.password_field :password,
+              placeholder: 'Enter password',
+              required: true,
+              input_html: { autocomplete: 'current-password' }, class: 'form-control'
+            = link_to 'Forgot your password?', new_password_path(User.new), class: 'help-block in-gold-500'
```
Comment (za4emnik) [inline app/views/checkout_steps/quick_registration.haml]: ?

File: app/views/checkout_steps/quick_registration.haml [hunk 39]
```diff
@@ -0,0 +1,51 @@
...
+            %br/
+    .col-md-5.col-md-offset-1.mb-60
+      .text-center
+        %h3.general-subtitle
+          = t('quick_registration.registration.title')
+        = link_to user_facebook_omniauth_authorize_path, class: 'general-login-icon' do
+          %i.fa.fa-facebook-official
+        %p.general-login-text or
+      %form.general-form
+      = simple_form_for(User.new, defaults: { input_html: { class: 'general-form' }}, url: registration_path(User.new)) do |f|
```
Comment (za4emnik) [inline app/views/checkout_steps/quick_registration.haml]: ?

File: app/views/checkout_steps/quick_registration.haml [hunk 40]
```diff
@@ -0,0 +1,51 @@
...
+          %p.mb-5
+            = t('quick_registration.registration.form_title')
+          %p.general-info-text.font-12.mb-20
+            = t('quick_registration.registration.form_notice')
+          = f.email_field :email,
+            id: 'quick_registration_email',
+            placeholder: I18n.t('quick_registration.login.enter_email'),
+            required: true,
+            autofocus: true,
+            input_html: { autocomplete: 'email' }, class: 'form-control'
```
Comment (za4emnik) [inline app/views/checkout_steps/quick_registration.haml]: text?

File: config/environments/production.rb [hunk 41]
```diff
@@ -39,7 +40,19 @@
...
 
   # Store uploaded files on the local file system (see config/storage.yml for options)
-  config.active_storage.service = :local
+  config.active_storage.service = :amazon
+  config.action_mailer.default_url_options = { host: ENV['APPLICATION_HOST'] }
+  config.action_mailer.delivery_method = :smtp
+  # SMTP settings for gmail
+  config.action_mailer.smtp_settings = {
+    address:                  ENV['GMAIL_SMPT'],
+    port:                     587,
```
Comment (za4emnik) [inline config/environments/production.rb]: ENV

File: config/environments/production.rb [hunk 42]
```diff
@@ -39,7 +40,19 @@
...
+    domain:                   ENV['APPLICATION_DOMAIN'],
+    authentication:           "plain",
```
Comment (za4emnik) [inline config/environments/production.rb]: ENV

File: spec/controllers/checkout_steps_controller_spec.rb (hunk 8 above)
Comment (za4emnik) [inline spec/controllers/checkout_steps_controller_spec.rb]: CONST

File: spec/controllers/order_items_controller_spec.rb [hunk 43]
```diff
@@ -0,0 +1,175 @@
...
+        end
+      end
+    end
+  end
+
+  describe '.destroy' do
+    subject{ delete :destroy, params: params }
+    context 'when user is guest' do
+      let(:quantity) { '1' }
+      let(:book_id) { '1' }
```
Comment (za4emnik) [inline spec/controllers/order_items_controller_spec.rb]: use once

File: spec/controllers/order_items_controller_spec.rb [hunk 44]
```diff
@@ -0,0 +1,175 @@
...
+      end
+
+      it 'add items to session' do
+        expect{subject}.to change {controller.session[:guest_items][order_item.id.to_s]}.from('1').to('2')
+      end
+    end
+
+    context 'when user is guest' do
+
+      before do
```
Comment (za4emnik) [inline spec/controllers/order_items_controller_spec.rb]: one line

File: spec/controllers/registrations_controller_spec.rb [hunk 45]
```diff
@@ -0,0 +1,65 @@
+RSpec.describe Users::RegistrationsController, type: :controller do
+  describe '#create' do
+    let(:password) { SecureRandom.base64(10) }
+    let(:password_confirmation) { password }
+    let(:params) { { 'user': { 'email': email,
+                               'password': password,
+                               'password_confirmation': password_confirmation,
```
Comment (za4emnik) [inline spec/controllers/registrations_controller_spec.rb]: traits

File: spec/controllers/users_controller_spec.rb (hunk 17 above)
Comment (za4emnik) [reply spec/controllers/users_controller_spec.rb]: ?

File: spec/decorators/book_decorator_spec.rb [hunk 46]
```diff
@@ -0,0 +1,21 @@
...
+  it 'returns the displayed days' do
+    expect(BookDecorator::RAITING_STARS).to eq([5, 4, 3, 2, 1])
+  end
+
+  it 'returns the displayed days' do
+    expect(book.reviews_quantity).to eq(0)
+  end
+
+  it 'returns the displayed days' do
+    expect(book.authors_all).to eq(book.authors.map { |a| "#{a.first_name} #{a.last_name}" }.join(', '))
```
Comment (za4emnik) [inline spec/decorators/book_decorator_spec.rb]: ?

File: spec/rails_helper.rb [hunk 47]
```diff
@@ -0,0 +1,28 @@
+require 'spec_helper'
+ENV['RAILS_ENV'] ||= 'test'
+
+require File.expand_path('../config/environment', __dir__)
+abort('The Rails environment is running in production mode!') if Rails.env.production?
+require 'rspec/rails'
+require 'webdrivers'
```
Comment (za4emnik) [inline spec/rails_helper.rb]: move

File: spec/support/capybara.rb [hunk 48]
```diff
@@ -0,0 +1,5 @@
+require 'capybara/rails'
+require 'capybara/rspec'
+require 'rack_session_access/capybara'
+
+# Capybara.default_driver = :selenium_chrome
```
Comment (za4emnik) [inline spec/support/capybara.rb]: remove

File: app/helpers/quick_registration_helper.rb [hunk 49]
```diff
@@ -0,0 +1,5 @@
+module QuickRegistrationHelper
```
Comment (za4emnik) [inline app/helpers/quick_registration_helper.rb]: remove, please
move to controller

File: config/initializers/active_admin.rb [hunk 50]
```diff
@@ -0,0 +1,23 @@
+ActiveAdmin.setup do |config|
+  # == Site Title
+  config.site_title = "Book Store"
```
Comment (za4emnik) [inline config/initializers/active_admin.rb]: try move to I18n

**Pull request merged**